src/builder/QuarkOperationHelper.sol

```diff
@@ -0,0 +1,150 @@
+            groupedQuarkOperations[i] = new IQuarkWallet.QuarkOperation[](operationCounts[i]);
+            groupedActions[i] = new Actions.Action[](operationCounts[i]);
+        }
+
+        // Group operations by chain
+        uint256[] memory currentIndex = new uint256[](uniqueChainCount);
+        for (uint256 i = 0; i < quarkOperations.length; ++i) {
+            uint256 chainId = actions[i].chainId;
+            for (uint256 j = 0; j < uniqueChainCount; j++) {
+                if (uniqueChainIds[j] == chainId) {
+                    groupedQuarkOperations[j][currentIndex[j]] = quarkOperations[i];
+                    groupedActions[j][currentIndex[j]] = actions[i];
+                    currentIndex[j]++;
+                    break;
+                }
+            }
+        }
+
+        // Create new arrays for merged operations and actions
+        IQuarkWallet.QuarkOperation[] memory mergedQuarkOperations = new IQuarkWallet.QuarkOperation[](uniqueChainCount);
+        Actions.Action[] memory mergedActions = new Actions.Action[](uniqueChainCount);
+
+        // Merge operations for each unique chain
+        for (uint256 i = 0; i < uniqueChainCount; ++i) {
+            if (operationCounts[i] == 1) {
+                // If there's only one operation for this chain, we don't need to merge
+                mergedQuarkOperations[i] = groupedQuarkOperations[i][0];
+                mergedActions[i] = groupedActions[i][0];
+            } else {
+                // Merge multiple operations for this chain
+                (mergedQuarkOperations[i], mergedActions[i]) =
+                    mergeOperations(groupedQuarkOperations[i], groupedActions[i]);
+            }
+        }
+
+        return (mergedQuarkOperations, mergedActions);
+    }
+
+    // Note: Assumes all the quark operations are for the same quark wallet.
+    function mergeOperations(IQuarkWallet.QuarkOperation[] memory quarkOperations, Actions.Action[] memory actions)
+        internal
+        pure
+        returns (IQuarkWallet.QuarkOperation memory, Actions.Action memory)
+    {
+        address[] memory callContracts = new address[](quarkOperations.length);
+        bytes[] memory callDatas = new bytes[](quarkOperations.length);
+        bytes[] memory scriptSources = new bytes[](quarkOperations.length + 1);
+
+        for (uint256 i = 0; i < quarkOperations.length; ++i) {
+            callContracts[i] = quarkOperations[i].scriptAddress;
+            callDatas[i] = quarkOperations[i].scriptCalldata;
+            // Note: For simplicity, we assume there is one script source per quark operation for now
+            scriptSources[i] = quarkOperations[i].scriptSources[0];
+        }
+        scriptSources[quarkOperations.length] = type(Multicall).creationCode;
+
+        bytes memory multicallCalldata = abi.encodeWithSelector(Multicall.run.selector, callContracts, callDatas);
+
+        // Construct QuarkOperation and Action
```

Comment:
We don't actually construct an Action here, right?

Reply:
Yeah we aren't constructing it, but we are picking an action (the last one) to return. Not sure what's a better way to convey it